### DISCUSSION TITLE
monitoring metric and model saving

### DISCUSSION THREAD

**Discussion by Struggle-Forever** [Lightning-AI/pytorch-lightning#13633] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
In pytorch-lightning,  we often monitor the metric at the current batch level，**validation_step**. Does this mean that the model parameters we save are optimal for the current batch rather than the whole validation set? Does this mean that in order to get the best model for the validation set, I have to monitor the metric in     **validation_epoch_end**?

```
    def share_val_step(self, batch):
        graph_label = batch.y
        graph_pred = self(data=batch)
        y_pred_label = graph_pred.argmax(dim=-1)
        f1, p, r, acc = envaulation(y_true=graph_label, y_pred=y_pred_label)
        return torch.tensor(acc, device=self.device), torch.tensor(f1, device=self.device)

    def validation_step(self, batch, batch_idx):
        f1, acc = self.share_val_step(batch)
        metrics = {'val_f1': f1, 'val_acc': acc}
        self.log_dict(metrics, prog_bar=True, logger=True, on_epoch=True)

    def validation_step_end(self, val_step_outputs):
        pass

    def validation_epoch_end(self, validation_step_outputs):
        pass

    def test_step(self, batch, batch_idx, dataloader_idx):
        f1, acc = self.share_val_step(batch)
        metrics = {'test_f1': f1, 'test_acc': acc}
        self.log_dict(metrics, prog_bar=True, logger=True, on_epoch=True)

```

**Comment by rohitgr7** [comment, marked as answer]
```py
f1, acc = self.share_val_step(batch)
metrics = {'val_f1': f1, 'val_acc': acc}
self.log_dict(metrics, prog_bar=True, logger=True, on_epoch=True)
```
here you are using `on_epoch=True`, which means the metric will be aggregated across the validation set and will be used to monitor the checkpoints if you set `ModelCheckpoint(..., monitor='val_acc')`.

**Reply by Struggle-Forever** [reply]
thank you very much! I like the pytorch-lightning!!!
